File: src/EggLedger.Web/wwwroot/js/blobcrypto.js
// AES-256-GCM blob crypto via the browser SubtleCrypto API. The WASM runtime
// does not support System.Security.Cryptography.AesGcm, so the Blazor client
// calls these instead. Wire format matches the Go server and the managed
// BlobCrypto: base64( nonce(12) || ciphertext || tag(16) ). SubtleCrypto's
// AES-GCM output is ciphertext||tag, identical to Go's gcm.Seal, so the layout
// lines up byte for byte.

const NONCE_SIZE = 12;

function hexToBytes(hex) {
  if (hex.length !== 64) {
    throw new Error(`AES-256 key must be 64 hex chars, got ${hex.length}`);
  }
  const out = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    out[i] = parseInt(hex.substr(i * 2, 2), 16);
  }
  return out;
}

function b64ToBytes(b64) {
  const bin = atob(b64);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) {
    out[i] = bin.charCodeAt(i);
  }
  return out;
}

function bytesToB64(bytes) {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) {
    bin += String.fromCharCode(bytes[i]);
  }
  return btoa(bin);
}

async function importKey(hexKey) {
  return crypto.subtle.importKey(
    "raw",
    hexToBytes(hexKey),
    { name: "AES-GCM" },
    false,
    ["encrypt", "decrypt"],
  );
}

// plaintextB64: base64 of the raw plaintext bytes (Blazor passes a byte[] as base64).
// returns base64( nonce || ciphertext || tag ).
export async function encrypt(hexKey, plaintextB64) {
  const key = await importKey(hexKey);
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_SIZE));
  const plaintext = b64ToBytes(plaintextB64);
  const sealed = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce, tagLength: 128 }, key, plaintext),
  );
  const out = new Uint8Array(NONCE_SIZE + sealed.length);
  out.set(nonce, 0);
  out.set(sealed, NONCE_SIZE);
  return bytesToB64(out);
}

// b64: base64( nonce || ciphertext || tag ). returns base64 of the plaintext bytes.
export async function decrypt(hexKey, b64) {
  const key = await importKey(hexKey);
  const data = b64ToBytes(b64);
  if (data.length < NONCE_SIZE + 16) {
    throw new Error("blob too short");
  }
  const nonce = data.slice(0, NONCE_SIZE);
  const sealed = data.slice(NONCE_SIZE); // ciphertext || tag
  const plaintext = new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce, tagLength: 128 }, key, sealed),
  );
  return bytesToB64(plaintext);
}

File: src/EggLedger.Domain/Crypto/BlobCrypto.cs
using System.Security.Cryptography;

namespace EggLedger.Domain.Crypto;

/// <summary>
/// AES-256-GCM blob encryption matching Go cloudsync.encryptBlob/decryptBlob. Wire format:
/// base64(12-byte nonce || ciphertext || 16-byte tag). Go appends the tag to the ciphertext while
/// .NET <see cref="AesGcm"/> takes it separately, so decrypt splits the trailing 16 bytes off.
/// </summary>
public static class BlobCrypto {
    private const int NonceSize = 12;
    private const int TagSize = 16;
    private const int KeySize = 32;

    /// <summary>
    /// Encrypts <paramref name="plaintext"/> with AES-256-GCM under the 64-char
    /// hex key. Returns base64(nonce || ciphertext || tag). Port of Go encryptBlob.
    /// </summary>
    public static string Encrypt(string hexKey, byte[] plaintext) {
        ArgumentNullException.ThrowIfNull(plaintext);
        var key = DecodeKey(hexKey);

        var nonce = new byte[NonceSize];
        RandomNumberGenerator.Fill(nonce);

        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[TagSize];
        using (var gcm = new AesGcm(key, TagSize)) {
            gcm.Encrypt(nonce, plaintext, ciphertext, tag);
        }

        // Lay out nonce || ciphertext || tag to match Go gcm.Seal(nonce, ...).
        var sealed_ = new byte[NonceSize + ciphertext.Length + TagSize];
        Buffer.BlockCopy(nonce, 0, sealed_, 0, NonceSize);
        Buffer.BlockCopy(ciphertext, 0, sealed_, NonceSize, ciphertext.Length);
        Buffer.BlockCopy(tag, 0, sealed_, NonceSize + ciphertext.Length, TagSize);
        return Convert.ToBase64String(sealed_);
    }

    /// <summary>
    /// Decrypts a base64(nonce || ciphertext || tag) blob produced by
    /// <see cref="Encrypt"/> or Go encryptBlob. Port of Go decryptBlob.
    /// </summary>
    public static byte[] Decrypt(string hexKey, string b64) {
        ArgumentNullException.ThrowIfNull(b64);
        var key = DecodeKey(hexKey);

        var data = Convert.FromBase64String(b64);
        if (data.Length < NonceSize + TagSize) {
            throw new CryptographicException("blob too short");
        }

        var nonce = new byte[NonceSize];
        Buffer.BlockCopy(data, 0, nonce, 0, NonceSize);

        var bodyLen = data.Length - NonceSize - TagSize;
        var ciphertext = new byte[bodyLen];
        Buffer.BlockCopy(data, NonceSize, ciphertext, 0, bodyLen);

        var tag = new byte[TagSize];
        Buffer.BlockCopy(data, NonceSize + bodyLen, tag, 0, TagSize);

        var plaintext = new byte[bodyLen];
        using (var gcm = new AesGcm(key, TagSize)) {
            gcm.Decrypt(nonce, ciphertext, tag, plaintext);
        }
        return plaintext;
    }

    private static byte[] DecodeKey(string hexKey) {
        ArgumentNullException.ThrowIfNull(hexKey);
        var key = Convert.FromHexString(hexKey);
        if (key.Length != KeySize) {
            throw new CryptographicException(
                $"AES-256 key must be {KeySize} bytes ({KeySize * 2} hex chars), got {key.Length}");
        }
        return key;
    }
}
